internal/graph/check.go

```diff
@@ -614,6 +614,241 @@ func (c *LocalChecker) ResolveCheck(
+	// [group:1#member, group:2#member, group:1#owner, group:3#owner] will be stored as
+	// [group#member][1]
+	// [group#owner][1, 3]
+	usersetsMap := make(map[string]map[string]struct{})
+
+	for {
+		// NOTE: For the future, once we observe a new ObjectRelation in the usersetsMap that means that all objectIDs
+		// needed for the intersection lookup for that ObjectRelation are in memory. There is no need to build the full
+		// usersetMap by draining the full iterator, it can start processing in batches of ObjectRelation.
+		// Consider doing it when we have better concurrency tooling for "streaming" operations.
+		t, err := iter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+			return nil, err
+		}
+
+		condEvalResult, err := eval.EvaluateTupleCondition(ctx, t, typesys, req.GetContext())
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		if len(condEvalResult.MissingParameters) > 0 {
+			errs = errors.Join(errs, condition.NewEvaluationError(
+				t.GetCondition().GetName(),
+				fmt.Errorf("tuple '%s' is missing context parameters '%v'",
+					tuple.TupleKeyToString(t),
+					condEvalResult.MissingParameters),
+			))
+			continue
+		}
+
+		if !condEvalResult.ConditionMet {
+			continue
+		}
+
+		object, relation := tuple.SplitObjectRelation(t.GetUser())
+		objectType, objectID := tuple.SplitObject(object)
+		objectRel := tuple.ToObjectRelationString(objectType, relation)
+		if _, ok := usersetsMap[objectRel]; !ok {
+			usersetsMap[objectRel] = make(map[string]struct{})
+		}
+		if _, ok := usersetsMap[objectRel][objectID]; !ok {
+			usersetsMap[objectRel][objectID] = struct{}{}
+		}
+	}
+
+	if errs != nil {
+		telemetry.TraceError(span, errs)
+		return nil, errs
+	}
+
+	storeID := req.GetStoreID()
+	reqTupleKey := req.GetTupleKey()
+
+	// Next, for all the ObjectRelation, compare the associated objectIDs
+	// to the users associated objects
+	// all of this can likely bee its own function
+	handlers := make([]CheckHandlerFunc, 0, len(usersetsMap))
+	for objectRel, objectIDs := range usersetsMap {
+		handler := func(ctx context.Context) (*ResolveCheckResponse, error) {
+			response := &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolveCheckResponseMetadata{
+					DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 1,
+				},
+			}
+			objectType, relation := tuple.SplitObjectRelation(objectRel)
+			iter, err := ds.ReadStartingWithUser(ctx, storeID, storage.ReadStartingWithUserFilter{
+				ObjectType: objectType,
+				Relation:   relation,
+				UserFilter: []*openfgav1.ObjectRelation{{
+					Object: reqTupleKey.GetUser(),
+				}},
+			})
+			if err != nil {
+				return nil, err
+			}
+			defer iter.Stop()
+
+			for {
+				t, err := iter.Next(ctx)
+				if err != nil {
+					if errors.Is(err, storage.ErrIteratorDone) {
+						break
+					}
+
+					return nil, err
+				}
+
+				_, objectID := tuple.SplitObject(t.GetKey().GetObject())
```

Comment:
don't we also have to evaluateTupleCondition here? 🤔  (how is this passing tests?)

Reply:
Why? The condition being evaluated at the source would discard it for the intersection 🤔

Reply:
But you could have both ends of the intersection be conditioned, e.g.

```go
doc:1, viewer, group:eng#member, cond: cond_x
user:maria, member, group:eng, cond:cond_x
```